### PR TITLE
Sample weights

### PR DESCRIPTION
**Pull request by nikisix**
Commit corresponds to discussion at https://github.com/maciejkula/spotlight/issues/118
Loss functions and implicit factorization done.
Implicit sequence coming.

### CODE REVIEW

File: spotlight/losses.py [hunk 1]
```diff
@@ -116,22 +157,22 @@ def hinge_loss(positive_predictions, negative_predictions, mask=None):
                        positive_predictions +
                        1.0, 0.0)
 
-    if mask is not None:
-        mask = mask.float()
-        loss = loss * mask
-        return loss.sum() / mask.sum()
+    if sample_weights is not None or mask is not None:
```
Comment (maciejkula) [inline spotlight/losses.py]: We don't need these checks here: we can always call the base loss function. I think that will make for less code repetition in all the loss functions.

File: spotlight/losses.py [hunk 2]
```diff
@@ -15,7 +15,42 @@
 from spotlight.torch_utils import assert_no_grad
 
 
-def pointwise_loss(positive_predictions, negative_predictions, mask=None):
+def base_loss(loss, sample_weights=None, mask=None):
```
Comment (maciejkula) [inline spotlight/losses.py]: Could we call this `_weighted_loss`? This would reflect the fact that:
1. This is an internal function.
2. Its essence lies in applying weights.

File: spotlight/factorization/implicit.py [hunk 3]
```diff
@@ -220,11 +219,114 @@ def fit(self, interactions, verbose=False):
...
+            epoch_loss /= minibatch_num + 1
+
+            if verbose:
+                print('Epoch {}: loss {}'.format(epoch_num, epoch_loss))
+
+            if np.isnan(epoch_loss) or epoch_loss == 0.0:
+                raise ValueError('Degenerate epoch loss: {}'
+                                 .format(epoch_loss))
+
+    def fit_weighted(self, interactions, verbose=False):
```
Comment (maciejkula) [inline spotlight/factorization/implicit.py]: I think we should keep a single `fit` method. If weights are present, we use them; if not, we don't.

This could probably be accomplished by changing the `minibatch` function to just infinitely yield `None` for those arguments that are `None` rather than tensors. This way, it always yields `None` for `batch_sample_weights` which ties in nicely with how the loss functions are changed.

File: spotlight/losses.py (hunk 2 above)
Comment (nikisix) [reply spotlight/losses.py]: I'm not opposed to this name, but it'd be a misnomer in case they call us with a mask and no weights. Although, conceptually at least, a mask could be thought of as a weight of zero.

_base_loss
vs
_weighted_loss
vs
_modified_loss

¯\_(ツ)_/¯ ?

File: spotlight/factorization/implicit.py (hunk 3 above)
Comment (nikisix) [reply spotlight/factorization/implicit.py]: Definitely tried to keep it to a single fit function at first. But I want to avoid consuming ram for an unnecessary tensor of Nones.
Would this lead to a tensor of Nones (not None) being passed to the loss functions, rather than a single None? (unless the tensor is not created by yielding Nones?)
What if we just checked if interactions.weights is specified at the beginning of fit, and if they are just have fit call _fit_weighted?

Sequence weight question for you: I was thinking about the sequence interactions yesterday, and my guess is to make a parallel sequence tensor full of sample weights, of the same dimensions as the sequence tensor of item-ids. Does this sound right?

File: spotlight/losses.py (hunk 1 above)
Comment (nikisix) [reply spotlight/losses.py]: totally

**Comment by maciejkula** [comment]
This is a good point: we can definitely subsume masks under weights by just setting weights to zero where mask should be false.

File: spotlight/factorization/implicit.py [hunk 4]
```diff
@@ -236,7 +247,106 @@ def fit(self, interactions, verbose=False):
...
+            epoch_loss /= minibatch_num + 1
+
+            if verbose:
+                print('Epoch {}: loss {}'.format(epoch_num, epoch_loss))
+
+            if np.isnan(epoch_loss) or epoch_loss == 0.0:
+                raise ValueError('Degenerate epoch loss: {}'
+                                 .format(epoch_loss))
+
+    def _fit_weighted(self, interactions, verbose=False):
```
Comment (maciejkula) [inline spotlight/factorization/implicit.py]: I'm still not convinced about having a separate `_fit_weighted` function, even if it is internal. I think it introduces too much code duplication that will have to be kept in sync.

What about modifying `minibatch` to look roughly like this:
```python
def minibatch(*tensors, **kwargs):

    batch_size = kwargs.get('batch_size', 128)

    if len(tensors) == 1:
        tensor = tensors[0]
        for i in range(0, len(tensor), batch_size):
            yield tensor[i:i + batch_size]
    else:
        for i in range(0, len(tensors[0]), batch_size):
            yield tuple(x[i:i + batch_size] if x is not None else None for x in tensors)
```
This way, it emits tensor slices if an argument is a tensor (as before), but also emits `None` in the tuple if an argument is `None`.

File: spotlight/factorization/implicit.py (hunk 3 above)
Comment (maciejkula) [reply spotlight/factorization/implicit.py]: No, we won't need a tensor of nones. I'll add a comment with a prototype of what I meant.

File: spotlight/losses.py (hunk 2 above)
Comment (maciejkula) [reply spotlight/losses.py]: If, as you suggest, we subsume masks under weights this will all be fine!
